File: perf/benchmarks/apps/word_count/word_count.cpp
// Word-count — typical text-processing application benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <unordered_map>

static std::string gen_text(int64_t n) {
    int64_t seed = 12345;
    std::string s;
    s.reserve(static_cast<size_t>(n) * 7);
    char buf[16];
    for (int64_t i = 0; i < n; ++i) {
        seed = (seed * 1103515245LL + 12345LL) & 0x7FFFFFFFLL;
        int idx = static_cast<int>(seed % 1024);
        std::snprintf(buf, sizeof(buf), "w%04d", idx);
        s.append(buf);
        if ((i % 12) == 11) s.push_back('\n');
        else                s.push_back(' ');
    }
    return s;
}

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 1'000'000LL;
    std::string text = gen_text(n);

    auto t0 = std::chrono::steady_clock::now();
    std::unordered_map<std::string, int64_t> counts;
    counts.reserve(2048);
    size_t i = 0, sz = text.size();
    while (i < sz) {
        // Skip whitespace
        while (i < sz && (text[i] == ' ' || text[i] == '\n')) ++i;
        size_t start = i;
        while (i < sz && text[i] != ' ' && text[i] != '\n') ++i;
        if (i > start) {
            ++counts[text.substr(start, i - start)];
        }
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    int64_t cs = 0;
    for (const auto& kv : counts) {
        // Parse digits after leading 'w'.
        int idx = std::atoi(kv.first.c_str() + 1);
        cs ^= static_cast<int64_t>(idx) * 1000003LL + kv.second;
    }
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/apps/word_count/word_count.py
# Word-count — typical text-processing application benchmark.
#
# Synthesizes a deterministic large body of text (LCG-generated word ids
# joined with spaces and newlines), then tokenizes it and counts word
# frequencies in a hash-map. Stresses string slicing/comparison, hashing,
# and hash-map updates.
#
# Argument: NUM_WORDS (default 1_000_000).

import sys
import time


def gen_text(n: int) -> str:
    seed = 12345
    parts = []
    for i in range(n):
        seed = (seed * 1103515245 + 12345) & 0x7FFFFFFF
        # 1024 distinct words via a small wordlist of 5-letter strings.
        idx = seed % 1024
        parts.append(f"w{idx:04d}")
        if (i % 12) == 11:
            parts.append("\n")
        else:
            parts.append(" ")
    return "".join(parts)


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 1_000_000
    text = gen_text(n)

    t0 = time.perf_counter_ns()
    counts: dict[str, int] = {}
    for w in text.split():
        counts[w] = counts.get(w, 0) + 1
    elapsed_ns = time.perf_counter_ns() - t0

    # Deterministic checksum: sum( hash-stable per word ) over distinct keys.
    cs = 0
    for k, v in counts.items():
        # Fold the digits of k (e.g. "w0427") into an integer.
        idx = int(k[1:])
        cs ^= idx * 1000003 + v
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
